vdirsyncer/storage/ldap.py

```diff
@@ -0,0 +1,118 @@
+            self.conn.bind()
+            self.conn.start_tls()
+
+        ldap_logger.debug('Connected to: {}'.format(self.conn))
+
+        if self.search_base is None:
+            # Fallback to default root entry
+            self.search_base = server.info.naming_contexts[0]
+
+    def list(self):
+        '''
+        :returns: list of (href, etag)
+        '''
+        ldap_logger.debug('Search on {self.search_base} with filter'
+                          '{self.filter}'.format(self=self))
+        self.conn.search(self.search_base, self.filter,
+                         attributes=["whenChanged"])
+        for entry in self.conn.entries:
+            ldap_logger.debug('Found {}'.format(entry.entry_get_dn()))
+            href = entry.entry_get_dn()
+            if getattr(entry, 'whenChanged'):
+                etag = str(entry.whenChanged)
+            else:
+                item = self.get(href)
+                etag = item.hash
+
+            yield href, etag
+
+    def get(self, href):
+        self.conn.search(href, self.filter,
+                         attributes=["whenChanged", "cn", "sn", "givenName",
+                                     "displayName", "telephoneNumber",
+                                     "mobile", "facsimileTelephoneNumber",
+                                     "mail", "title"])
+
+        if not self.conn.entries[0]:
+            raise exceptions.NotFoundError(href)
+
+        entry = self.conn.entries[0]
+        etag = str(entry.whenChanged)
+
+        vcard = vobject.vCard()
+        vo = vcard.add('fn')
+        vo.value = str(entry.cn)
+        vo = vcard.add('n')
+        vo.value = vobject.vcard.Name(family=str(entry.sn),
+                                      given=str(entry.givenName))
+        if getattr(entry, 'telephoneNumber', None):
+            vo = vcard.add('tel')
+            vo.value = str(entry.telephoneNumber)
+            vo.type_param = 'WORK'
+        if getattr(entry, 'mobile', None):
+            vo = vcard.add('tel')
+            vo.value = str(entry.mobile)
+            vo.type_param = 'CELL'
+        if getattr(entry, 'facsimileTelephoneNumber', None):
+            vo = vcard.add('tel')
+            vo.value = str(entry.facsimileTelephoneNumber)
+            vo.type_param = 'FAX'
+        if getattr(entry, 'mail', None):
+            vo = vcard.add('email')
+            vo.value = str(entry.mail)
+            vo.type_param = 'INTERNET'
+        if getattr(entry, 'title', None):
+            vo = vcard.add('title')
+            vo.value = str(entry.title)
+
+        item = Item(vcard.serialize())
+
+        return item, etag
+
+    def upload(self, item):
+        vcard = vobject.readOne(item.raw)
+        self.conn.strategy.add_entry('cn={},ou=test,o=lab'.format(vcard.fn),
+                                     vcard)
+
+    def update(self, href, item, etag):
+        vcard = vobject.readOne(item.raw)
+        self.conn.strategy.add_entry(href, vcard)
```

Comment:
`add_entry` will do nothing if the item already exists, but if I call `update`, the item always exists. `update` is also expected to fail if the passed `etag` does not match the `etag` of the item on the server.

I think `modify` or how it's called is more appropriate than `add_entry`, but I'm not sure since I couldn't find much docs :(